lib/rules/color-hex-case/index.js

```diff
@@ -26,68 +30,60 @@ function rule(expectation, options, context) {
 		}
 
 		root.walkDecls((decl) => {
-			const declString = blurFunctionArguments(decl.toString(), 'url');
-			const fixPositions = [];
+			const parsedValue = valueParser(getValue(decl));
+			let needsFix = false;
 
-			styleSearch({ source: declString, target: '#' }, (match) => {
-				const hexMatch = /^#[0-9A-Za-z]+/.exec(declString.substr(match.startIndex));
+			parsedValue.walk((node) => {
+				const { value } = node;
 
-				if (!hexMatch) {
-					return;
-				}
+				if (isIgnoredFunction(node)) return false;
 
-				const hexValue = hexMatch[0];
-				const hexValueLower = hexValue.toLowerCase();
-				const hexValueUpper = hexValue.toUpperCase();
-				const expectedHex = expectation === 'lower' ? hexValueLower : hexValueUpper;
+				if (!isHexColor(node)) return;
 
-				if (hexValue === expectedHex) {
-					return;
-				}
+				const expected = expectation === 'lower' ? value.toLowerCase() : value.toUpperCase();
+
+				if (value === expected) return;
 
 				if (context.fix) {
-					fixPositions.unshift({
-						expectedHex,
-						currentHex: hexValue,
-						startIndex: match.startIndex,
-					});
+					node.value = expected;
+					needsFix = true;
 
 					return;
 				}
 
 				report({
-					message: messages.expected(hexValue, expectedHex),
+					message: messages.expected(value, expected),
 					node: decl,
-					index: match.startIndex,
+					index: declarationValueIndex(decl) + node.sourceIndex,
 					result,
 					ruleName,
 				});
 			});
 
-			if (fixPositions.length) {
-				const declProp = decl.prop;
-				const declBetween = decl.raws.between;
-
-				fixPositions.forEach((fixPosition) => {
-					// 1 — it's a # length
-					decl.value = replaceHex(
-						decl.value,
-						fixPosition.currentHex,
-						fixPosition.expectedHex,
-						fixPosition.startIndex - declProp.length - declBetween.length - 1,
-					);
-				});
+			if (needsFix) {
+				setValue(decl, parsedValue.toString());
 			}
 		});
 	};
 }
 
-function replaceHex(input, searchString, replaceString, startIndex) {
-	const offset = startIndex + 1;
-	const stringStart = input.slice(0, offset);
-	const stringEnd = input.slice(offset + searchString.length);
+function isIgnoredFunction({ type, value }) {
+	return type === 'function' && IGNORED_FUNCTIONS.has(value.toLowerCase());
+}
+
+function isHexColor({ type, value }) {
+	return type === 'word' && HEX.test(value);
+}
+
+function getValue(decl) {
+	return decl.raws.value ? decl.raws.value.raw : decl.value;
+}
+
+function setValue(decl, value) {
+	if (decl.raws.value) decl.raws.value.raw = value;
+	else decl.value = value;
 
-	return stringStart + replaceString + stringEnd;
+	return decl;
```

Comment:
Return is not used anywhere.

Reply:
@Dru89 It looks like this is the only change we need to make for a second approval.

Reply:
@Dru89 It looks like this is the only change we need to make for a second approval.